src/main/scala/firrtl/passes/NewLowerTypes.scala

```diff
@@ -0,0 +1,439 @@
+      val loweredClock = onExpression(clock)
+      val loweredReset = onExpression(reset)
+      // It is important to first lower the declaration, because the reset can refer to the register itself!
+      val loweredRegs = symbols.lower(d.name, d.tpe, firrtl.RegKind)
+      val inits = Utils.create_exps(d.init).map(onExpression)
+      Block(
+        loweredRegs.zip(inits).map { case ((name, tpe, _), init) =>
+          DefRegister(info, name, tpe, loweredClock, loweredReset, init)
+      })
+    case d : DefNode =>
+      val values = Utils.create_exps(d.value).map(onExpression)
+      Block(
+        symbols.lower(d.name, d.value.tpe, firrtl.NodeKind).zip(values).map{ case((name, tpe, _), value) =>
+          assert(tpe == value.tpe)
+          DefNode(d.info, name, value)
+      })
+    case d : DefMemory =>
+      // TODO: as an optimization, we could just skip ground type memories here.
+      //       This would require that we don't error in getReferences() but instead return the old reference.
+      val mems = symbols.lower(d)
+      if(mems.length > 1 && memInit.exists(_.target.ref == d.name)) {
+        val mod = memInit.find(_.target.ref == d.name).get.target.encapsulatingModule
+        val msg = s"[module $mod] Cannot initialize memory ${d.name} of non ground type ${d.dataType.serialize}"
+        throw new RuntimeException(msg)
+      }
+      Block(symbols.lower(d))
+    case d : DefInstance => symbols.lower(d)
+    // connections
+    case Connect(info, loc, expr) =>
+      if(!expr.tpe.isInstanceOf[GroundType]) {
+        throw new RuntimeException(s"NewLowerTypes expects Connects to have been expanded! ${expr.tpe.serialize}")
+      }
+      val rhs = onExpression(expr)
+      // We can get multiple refs on the lhs because of ground-type memory ports like "clk" which can get duplicated.
+      val lhs = symbols.getReferences(loc.asInstanceOf[RefLikeExpression])
+      Block(lhs.map(loc => Connect(info, loc, rhs)))
+    case p : PartialConnect =>
+      throw new RuntimeException(s"NewLowerTypes expects PartialConnects to be resolved! $p")
+    case IsInvalid(info, expr) =>
+      if(!expr.tpe.isInstanceOf[GroundType]) {
+        throw new RuntimeException(s"NewLowerTypes expects IsInvalids to have been expanded! ${expr.tpe.serialize}")
+      }
+      // We can get multiple refs on the lhs because of ground-type memory ports like "clk" which can get duplicated.
+      val lhs = symbols.getReferences(expr.asInstanceOf[RefLikeExpression])
+      Block(lhs.map(loc => IsInvalid(info, loc)))
+    // others
+    case other => other.mapExpr(onExpression).mapStmt(onStatement)
+  }
+
+  /** Replaces all Reference, SubIndex and SubField nodes with the updated references */
+  def onExpression(e: Expression)(implicit symbols: LoweringTable): Expression = e match {
+    case r: RefLikeExpression =>
+      // When reading (and not assigning to) an expression, we can always just pick the first one.
+      symbols.getReferences(r).head
```

Comment:
Can you update the comment to give an example of when getReferences returns a list of size greater than one? I imagine the memory's clock case is one example.